### ISSUE TITLE
[Compat] Armourer's Workshop support?

### ISSUE THREAD

**Issue by DoomRater** [Elveskevtar/SmartMoving#6]
After giving the current build a go I can see that Armourer's Workshop isn't quite compatible with RenderAPI just yet.  I'm not sure whose end work needs to be done on, but things actually work a bit better than I expected.  For example, arms rotate upwards correctly when flying or swimming, but AW cosmetics don't follow the rotation of the player at all.  It doesn't look like much needs to be fixed here for the two to be compatible.

(edit: my observations are based on the wrong SmartMoving build! will repost when I actually use the build from GitHub.)

My hope is that the two mods will end up compatible with each other down the road.  Most mods don't even render the AW cosmetics at all, but because this uses a common enough library...

**Comment by DoomRater** [comment]
Just tried with the 17.0RC1 and one rotation seems to be off with Armourer's Workshop.  I'll try to take a look and see where the issue lies, whether with one of the APIs this mod uses or with AW.

**Comment by DoomRater** [comment]
I also tested with Extra Bit Manipulation, an addon for Chisels and Bits that lets you wear specially customized armors, and it too fails to see rotations.  Animations for different parts don't match up either.  I doubt this one will be compatible without the other mod doing something to compensate.

**Comment by elveskevtar** [comment]
Could you upload some pictures/videos of this? I'm assuming these are things that you can add on to the player model from other mods. If so, things like armor, parrots, custom heads, etc. all have their own "custom layer." I had to make some changes to the mod to correctly do the rotations on this for even the vanilla stuff. It did not seem hacky but maybe there is a way to do this without having to specify every custom layer. It could really help with external mod support for mods that add on top of the main model and also for any similar things they add to vanilla MC in the future.

**Comment by DoomRater** [comment]
![2020-04-09_22 02 40](https://user-images.githubusercontent.com/4010432/79055521-c3d99800-7c1b-11ea-8082-4f0dcf9bf016.png)
This screenshot looks way more broken than it really is.  AW chest piece doesn't match the rotation of the player, but the arms and legs are following custom animations really closely.  My net's a little annoying so it'll be a moment before I can upload video.  Currently waiting on the other video to upload before I do another.
![2020-04-09_22 02 50](https://user-images.githubusercontent.com/4010432/79055587-2af74c80-7c1c-11ea-8583-119d2ad6a396.png)
It'd also help if I had worn pieces of AW armor that weren't skirts...

**Comment by elveskevtar** [comment]
Hmmm, interesting. It looks like the armor is somewhat following the player's arms when flying but I can't tell how exact because it looks a bit off in that picture. Not rush on the video but it would be helpful to see.

Out of curiosity, are there any headpieces in the mod? If so, how are the models following the SmartMoving head?

**Comment by DoomRater** [comment]
Right, lemme show off some cosmetics on head, chest, legs, and feet.  This uses the Doom Marine armor found in the Global Library and the space helmet, also found in the Global Library.
https://imgur.com/a/GjvVHv6
Any part still made out of the Steve model works fine, but any part made out of the AW parts is just locked towards south.

**Comment by elveskevtar** [comment]
It's hilarious how the arm and leg pieces to the correct small rotations in reference to where the original model would be put don't actually rotate with the whole rest of the player. Thank you, that helps a lot.

**Comment by EzerArch** [comment]
Hello, I'm a user of SmartMoving+Armourer's Workshop in MC 1.7.10 ([video](https://youtu.be/kLYihlYpcks?t=278)). Thanks @Elveskevtar for continuing it!

Elaborating further on what @DoomRater has shown, it seems SM takes in account only the 1st layer of the player model but ignores the 2nd layer, be it the Minecraft overlay skin (which is not mod-dependent) or a high-poly skin such as Armourer's Workshop (Note: I'm assuming these two layers exist).

I've recorded a short video demonstrating it: https://www.twitch.tv/videos/590429476
- 0:10 - 0:38 : I turned on the skin customization parts, which is built-in (not modded). You see these parts don't follow properly the rotations.
- 0:39 - 0:59 : I put on some Armourer's Workshop skins which are attacked directly to the player model. -- they are called skin paint and have zero thickness -- they follow the player properly.
- 1:10 - end : I added some Armourer's Workshop skins which are more detailed -- they have thickness -- they don't follow the player properly. (edit: if the video "skips" this part, please use the playhead)

The same issue may occur with other mods, as has happened in Minecraft 1.7.10 and earlier. If the mod has an armor set which "extends" from vanilla armor method, Smart Moving will work seamlessly. But if the mod has its own armor models, the armor will be "locked" toward the same angle.
